doc/src/content/motivation/narrow_contract.md

```diff
@@ -0,0 +1,124 @@
++++
+title = "Narrow contracts"
+description = "Describes narrow-contract functions that do not work for all input values, and advantage of using them."
+weight = 60
++++
+
+A program can enter a "disappointing" state for two reasons:
+
+* due to disappointing situation in the environment (operating system, external input),
+  or
+* due to a bug in the program.
+
+The key to handling these disappointments correctly is to identify to which
+category they belong, and use the tools adequate for a given category. In this
+tutorial when we say "error" or "failure" we only refer to the first category.
+A bug is not an error.
+
+A bug is when a program is something else than what it is supposed to be. The
+correct action in that case is to change the program so that it is exactly what
+it is supposed to be. Unfortunately, sometimes the symptoms of a bug are only
+detected when the system is running and at this point no code changes are possible.
+
+In contrast, a failure is when a correct function in a correct program reflects
+some disappointing behavior in the environment. The correct action in that case
+is for the program to take a control path different than usual, which will likely
+cancel some operations and will likely result in different communication with the
+outside world.
+
+Symptoms of bugs can sometimes be detected during compilation or static program
+analysis or at run-time when observing certain values of objects that are declared
+never to be valid at certain points. One classical example is passing a null pointer
+to functions that expect a pointer to a valid object:
+
+```
+int f(int * pi) // expects: pi != nullptr
+{
+  return *pi + 1;
+}
+```
+
+Passing a null pointer where it is not expected is so common a bug that tools
+are very good at finding them. For instance, static analyzers will usually detect
+it without even executing your code. Similarly, tools like undefined behavior
+sanitizers will compile a code as the one above so that a safety check is performed
+to check if the pointer is null, and an error message will be logged and program
+optionally terminated.
+
+More, compilers can perform optimizations based on undefined behavior caused by
+dereferencing a null pointer. In the following code:
+
+```
+pair<int, int> g(int * pi) // expects: pi != nullptr
+{
+  int i = *pi + 1;
+  int j = pi == nullptr ? 1 : 0;
+  return {i, j};
+}
+```
+
+The compiler can see that if `pi` is null, the program would have undefined
+behavior. Since undefined behavior is never the programmers intention, the compiler
```

Comment:
Since undefined behaviour is required by the C++ standard to be never the programmer's intention